### PR TITLE
Add jvm_thread metrics

### PR DESCRIPTION
**Pull request by stalary**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
https://github.com/apache/skywalking/issues/5098

___
### New feature or improvement
- Describe the details and related test reports.
Add jvm_thread metrics

### CODE REVIEW

**Comment by stalary** [comment]
I'm sorry that the previous PR was wrong due to my operation, which led to the error of the sub-module. So I started a PR again @wu-sheng

File: oap-server/server-bootstrap/src/main/resources/oal/java-agent.oal [hunk 1]
```diff
@@ -25,4 +25,7 @@ instance_jvm_memory_noheap_max = from(ServiceInstanceJVMMemory.max).filter(heapS
 instance_jvm_young_gc_time = from(ServiceInstanceJVMGC.time).filter(phrase == GCPhrase.NEW).sum();
 instance_jvm_old_gc_time = from(ServiceInstanceJVMGC.time).filter(phrase == GCPhrase.OLD).sum();
 instance_jvm_young_gc_count = from(ServiceInstanceJVMGC.count).filter(phrase == GCPhrase.NEW).sum();
-instance_jvm_old_gc_count = from(ServiceInstanceJVMGC.count).filter(phrase == GCPhrase.OLD).sum();
\ No newline at end of file
+instance_jvm_old_gc_count = from(ServiceInstanceJVMGC.count).filter(phrase == GCPhrase.OLD).sum();
+instance_jvm_thread_live_count = from(ServiceInstanceThreadGC.liveCount).sum();
+instance_jvm_thread_daemon_count = from(ServiceInstanceThreadGC.daemonCount).sum();
+instance_jvm_thread_peak_count = from(ServiceInstanceThreadGC.peakCount).sum();
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/oal/java-agent.oal:31]: Why do you `sum` the count? This value could be collected multiple times, then the thread number(sum value) would be very large.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/thread/ThreadProvider.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+public enum ThreadProvider {
+    INSTANCE;
+    private final ThreadMXBean threadMXBean;
+
+    ThreadProvider() {
+        this.threadMXBean = ManagementFactory.getThreadMXBean();
+    }
+
+    public Thread getThreadMetric() {
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/thread/ThreadProvider.java:33]: `getThreadMetric` -> `getThreadMetrics`.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ServiceInstanceJVMThread.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+import lombok.Getter;
+import lombok.Setter;
+
+import static org.apache.skywalking.oap.server.core.source.DefaultScopeDefine.SERVICE_INSTANCE_CATALOG_NAME;
+import static org.apache.skywalking.oap.server.core.source.DefaultScopeDefine.SERVICE_INSTANCE_JVM_THREAD;
+
+@ScopeDeclaration(id = SERVICE_INSTANCE_JVM_THREAD, name = "ServiceInstanceJVMThread", catalog = SERVICE_INSTANCE_CATALOG_NAME)
+@ScopeDefaultColumn.VirtualColumnDefinition(fieldName = "entityId", columnName = "entity_id", isID = true, type = String.class)
+public class ServiceInstanceJVMThread extends Source {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ServiceInstanceJVMThread.java:29]: As you added a new source, `scope-definitions.md` doc should be added.

File: oap-server/server-bootstrap/src/main/resources/oal/java-agent.oal (hunk 1 above)
Comment (stalary) [reply oap-server/server-bootstrap/src/main/resources/oal/java-agent.oal:31]: Sorry, I was wrong here

File: oap-server/server-bootstrap/src/main/resources/oal/java-agent.oal [hunk 4]
```diff
@@ -26,6 +26,6 @@ instance_jvm_young_gc_time = from(ServiceInstanceJVMGC.time).filter(phrase == GC
 instance_jvm_old_gc_time = from(ServiceInstanceJVMGC.time).filter(phrase == GCPhrase.OLD).sum();
 instance_jvm_young_gc_count = from(ServiceInstanceJVMGC.count).filter(phrase == GCPhrase.NEW).sum();
 instance_jvm_old_gc_count = from(ServiceInstanceJVMGC.count).filter(phrase == GCPhrase.OLD).sum();
-instance_jvm_thread_live_count = from(ServiceInstanceThreadGC.liveCount).sum();
-instance_jvm_thread_daemon_count = from(ServiceInstanceThreadGC.daemonCount).sum();
-instance_jvm_thread_peak_count = from(ServiceInstanceThreadGC.peakCount).sum();
\ No newline at end of file
+instance_jvm_thread_live_count = from(ServiceInstanceThreadGC.liveCount);
+instance_jvm_thread_daemon_count = from(ServiceInstanceThreadGC.daemonCount);
+instance_jvm_thread_peak_count = from(ServiceInstanceThreadGC.peakCount);
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/oal/java-agent.oal:31]: No function? It is strange.

Comment (stalary) [reply oap-server/server-bootstrap/src/main/resources/oal/java-agent.oal:31]: longAvg() looks suitable

**Comment by arugal** [comment]
[OALLexer.g4](https://github.com/apache/skywalking/blob/master/oap-server/oal-grammar/src/main/antlr4/org/apache/skywalking/oal/rt/grammar/OALLexer.g4) and [OALParser.g4](https://github.com/apache/skywalking/blob/master/oap-server/oal-grammar/src/main/antlr4/org/apache/skywalking/oal/rt/grammar/OALParser.g4) are missing, Please follow [scope-definitions.md](https://github.com/apache/skywalking/blob/master/docs/en/concepts-and-designs/scope-definitions.md).

**Comment by stalary** [comment]
> [OALLexer.g4](https://github.com/apache/skywalking/blob/master/oap-server/oal-grammar/src/main/antlr4/org/apache/skywalking/oal/rt/grammar/OALLexer.g4) and [OALParser.g4](https://github.com/apache/skywalking/blob/master/oap-server/oal-grammar/src/main/antlr4/org/apache/skywalking/oal/rt/grammar/OALParser.g4) are missing, Please follow [scope-definitions.md](https://github.com/apache/skywalking/blob/master/docs/en/concepts-and-designs/scope-definitions.md).

Thanks. I will correct it later

**Comment by arugal** [comment]
Can you add tests related to JVM metrics in E2E :)  In this [position](
https://github.com/apache/skywalking/blob/e0325a44070cf94c73624a2c4c81819bc14ee09d/test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/simple/SimpleE2E.java#L181).  FYI @kezhenxu94

**Comment by stalary** [comment]
> Can you add tests related to JVM metrics in E2E? In this [position](https://github.com/apache/skywalking/blob/e0325a44070cf94c73624a2c4c81819bc14ee09d/test/e2e/e2e-test/src/test/java/org/apache/skywalking/e2e/simple/SimpleE2E.java#L181). FYI @kezhenxu94

Ok, let me see here

**Comment by stalary** [comment]
I'm not expected/simple/serviceInstanceTopo yml found related to the JVM configuration, how should I go to modify? @arugal

**Comment by kezhenxu94** [comment]
@stalary 

- add a new test case to verify the JVM-related metrics
- add a expected file (`expected/simple/jvmThreadMetrics.yml`)

here is a skeleton

```java

    @RetryableTest
    void jvmThreadMetrics() throws Exception {
        // query the jvm thread metrics

        load("expected/simple/jvmThreadMetrics.yml").as(JvmThreadMetricsMatcher.class).verify(jvmThreadMetrics);
    }
```

I don't think it's possible to verify the exact thread count, active thread count, etc, so verifying that they are > 0 should be enough

**Comment by stalary** [comment]
thanks @kezhenxu94

**Comment by wu-sheng** [comment]
@stalary Please provide a screenshot for the new instance dashboard.

**Comment by wu-sheng** [comment]
@arugal Could you recheck this locally?

**Comment by stalary** [comment]
截屏2020-07-21 下午10 05 25

**Comment by wu-sheng** [comment]
Where are the metrics?

**Comment by stalary** [comment]
> Where are the metrics?

Just a moment, I haven't started my service yet

**Comment by stalary** [comment]
截屏2020-07-21 下午10 14 28
When will JVM data be triggered? I called the service, but none of the JVM data is available

**Comment by stalary** [comment]
Wait a minute. I found a problem with my local tests

**Comment by wu-sheng** [comment]
You are adding thread metrics, where are they? I think this is your PR for, isn't it?

**Comment by stalary** [comment]
`
ERROR 2020-07-21 22:21:06:778 SkywalkingAgent-6-JVMService-produce-0 JVMService : JVMService produces metrics failure.
java.lang.NoSuchMethodError: org.apache.skywalking.apm.network.language.agent.v3.Thread$Builder.setLiveCount(I)Lorg/apache/skywalking/apm/network/language/agent/v3/Thread$Builder;
	at org.apache.skywalking.apm.agent.core.jvm.thread.ThreadProvider.getThreadMetrics(ThreadProvider.java:37)
	at org.apache.skywalking.apm.agent.core.jvm.JVMService.run(JVMService.java:125)
	at org.apache.skywalking.apm.util.RunnableWithExceptionProtection.run(RunnableWithExceptionProtection.java:33)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
`

There's a problem. Let me see why

**Comment by stalary** [comment]
I use `./mvnw clean package -DskipTests`
It seems that The proto I modified has not been updated. Is my operation wrong? @wu-sheng

**Comment by wu-sheng** [comment]
Whether submodule updated or not, that depends on your local git status. Nothing related to maven command. Those are totally different.

**Comment by stalary** [comment]
I encountered this error, but I have no idea. Could you give me some help？
`
ERROR 2020-07-21 22:50:03:037 SkywalkingAgent-7-JVMService-consume-0 JVMService : send JVM metrics to Collector fail.
org.apache.skywalking.apm.dependencies.io.grpc.StatusRuntimeException: UNKNOWN
        at org.apache.skywalking.apm.dependencies.io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:240)
        at org.apache.skywalking.apm.dependencies.io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:221)
        at org.apache.skywalking.apm.dependencies.io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:140)
        at org.apache.skywalking.apm.network.language.agent.v3.JVMMetricReportServiceGrpc$JVMMetricReportServiceBlockingStub.collect(JVMMetricReportServiceGrpc.java:158)
        at org.apache.skywalking.apm.agent.core.jvm.JVMService$Sender.run(JVMService.java:155)
        at org.apache.skywalking.apm.util.RunnableWithExceptionProtection.run(RunnableWithExceptionProtection.java:33)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
`

**Comment by wu-sheng** [comment]
Whether submodule updated or not, that depends on your local git status. Nothing related to maven command. Those are totally different.

How do you test this in the first place? Your questions are confusing me.

**Comment by stalary** [comment]
> How do you test this in the first place? Your questions are confusing me.

I used `./mvnw package -DskipTests` in my code branch, then run oap-server and test it with the service configuration -javaagent

**Comment by wu-sheng** [comment]
I mean, you submitted this PR, but not tested locally?

**Comment by stalary** [comment]
Sorry, I was running a local test this morning, but I haven't finished it yet

**Comment by wu-sheng** [comment]
The graph type is not right. Take a look on the 2rd tool, it should be linear graph.

File: oap-server/server-bootstrap/src/main/resources/ui-initialized-templates.yml [hunk 5]
```diff
@@ -359,6 +359,17 @@ templates:
...
                 },
+                {
+                  "width": 3,
+                  "title": "JVM Thread Count (Java Service)",
+                  "height": "250",
+                  "entityType": "ServiceInstance",
+                  "independentSelector": false,
+                  "metricType": "REGULAR_VALUE",
+                  "queryMetricType": "readMetricsValues",
+                  "chartType": "ChartBar",
```
Comment (arugal) [inline oap-server/server-bootstrap/src/main/resources/ui-initialized-templates.yml:370]:
```suggestion
                  "chartType": "ChartLine",
```

Comment (arugal) [reply oap-server/server-bootstrap/src/main/resources/ui-initialized-templates.yml:370]: ### screenshots
image

**Comment by wu-sheng** [comment]
@stalary Every time you submit pull request, it should and must be tested locally.

**Comment by stalary** [comment]
> @stalary Every time you submit pull request, it should and must be tested locally.

I'm very sorry. I will pay attention to it next time

**Comment by stalary** [comment]
截屏2020-07-22 上午12 56 31
截屏2020-07-22 上午12 56 48

Above are my local test results

**Comment by mrproliu** [comment]
@stalary Hi, I see that you have changed the JVMMetrics.proto file, but I can't found it at the [apm-protocol project PR list](https://github.com/apache/skywalking-data-collect-protocol/pulls?q=is%3Apr+is%3Aclosed), Have you submitted PR to the main project? Do I miss anything?

**Comment by stalary** [comment]
https://github.com/apache/skywalking-data-collect-protocol/pull/28

**Comment by mrproliu** [comment]
> [apache/skywalking-data-collect-protocol#28](https://github.com/apache/skywalking-data-collect-protocol/pull/28) @mrproliu

Thanks, have merged. Please update the protocol submodule in the [main repo](https://github.com/apache/skywalking). prevent errors when others download the Skywalking project.

**Comment by stalary** [comment]
> > [apache/skywalking-data-collect-protocol#28](https://github.com/apache/skywalking-data-collect-protocol/pull/28) @mrproliu
> 
> Thanks, have merged. Please update the protocol submodule in the [main repo](https://github.com/apache/skywalking). prevent errors when others download the Skywalking project.

ok

**Comment by stalary** [comment]
https://github.com/apache/skywalking/pull/5161

**Pull request merged**